File: go/core/agent_test.go
package core

import (
	"context"
	"strings"
	"testing"
)

// These agent tests are driven by the reusable MockLlmProvider (see
// llm_provider.go) as the demonstration that it replaces the hand-written
// fakeClient. The shared fakeClient lives in testhelpers_test.go for the other
// per-feature suites that still use it.

func TestKnowledgeRanksByOverlap(t *testing.T) {
	kb := &InMemoryKnowledge{}
	kb.Ingest("The return window is 17 days from delivery.", "returns.md")
	kb.Ingest("Gift wrapping costs 4.99 per item.", "wrapping.md")
	hits := kb.Query("what is the return window?", 1)
	if len(hits) != 1 {
		t.Fatalf("want 1 hit, got %d", len(hits))
	}
	if !strings.Contains(hits[0].Content, "17 days") {
		t.Fatalf("top hit should be the returns doc, got %q", hits[0].Content)
	}
}

func TestTextReplyStopsAfterOneCall(t *testing.T) {
	mock := NewMockLlmProvider().PushText("the answer is 42")
	agent := NewSmoothAgent(mock, AgentOptions{Instructions: "be helpful"})
	res, err := agent.Run(context.Background(), "what is the answer?", nil)
	if err != nil {
		t.Fatal(err)
	}
	if res.Text != "the answer is 42" || res.Iterations != 1 || res.ToolCalls != 0 {
		t.Fatalf("unexpected result: %+v", res)
	}
}

func TestToolCallThenFinish(t *testing.T) {
	echo := FuncTool{
		ToolName: "echo",
		Desc:     "Echoes input back",
		Params:   map[string]any{"type": "object", "properties": map[string]any{"text": map[string]any{"type": "string"}}, "required": []string{"text"}},
		Fn: func(_ context.Context, args map[string]any) (string, error) {
			s, _ := args["text"].(string)
			return s, nil
		},
	}
	mock := NewMockLlmProvider()
	mock.PushToolCall("call-1", "echo", `{"text": "hello tools"}`).PushText("done")
	agent := NewSmoothAgent(mock, AgentOptions{Tools: []Tool{echo}})
	res, err := agent.Run(context.Background(), "use echo", nil)
	if err != nil {
		t.Fatal(err)
	}
	if res.Text != "done" || res.ToolCalls != 1 {
		t.Fatalf("unexpected result: %+v", res)
	}
	// The tool result was fed back as a tool-role message before the final call.
	second := mock.Calls()[1].Messages
	found := false
	for _, m := range second {
		if m.Role == "tool" && m.Content == "hello tools" {
			found = true
		}
	}
	if !found {
		t.Fatalf("tool result not fed back; messages=%+v", second)
	}
}

func TestKnowledgeInjectedIntoSystemPrompt(t *testing.T) {
	kb := &InMemoryKnowledge{}
	kb.Ingest("The return window is exactly 17 days from delivery.", "returns.md")
	mock := NewMockLlmProvider().PushText("17 days")
	agent := NewSmoothAgent(mock, AgentOptions{Instructions: "support agent", Knowledge: kb})
	if _, err := agent.Run(context.Background(), "how many days to return?", nil); err != nil {
		t.Fatal(err)
	}
	first := mock.Calls()[0].Messages
	if len(first) == 0 || first[0].Role != "system" || !strings.Contains(first[0].Content, "17 days") {
		t.Fatalf("knowledge not injected into system prompt; first=%+v", first)
	}
}

File: typescript/core/test/agent.test.ts
/**
 * Non-network unit tests for the TypeScript core: the agentic loop, tool calling,
 * and knowledge injection, driven by the reusable `MockLlmProvider`. Always green
 * (no credentials) — live-gateway behavior is covered by `evals.test.ts`.
 *
 * These tests used to roll their own ad-hoc `FakeClient`; they now use the shared
 * `MockLlmProvider` (see `llmProvider.test.ts`) as the demonstration that it
 * replaces the hand-written fakes.
 */

import { describe, expect, it } from 'vitest';
import { AgentOptions, effectiveMaxTokens, SmoothAgent, Tool } from '../src/agent.js';
import { InMemoryKnowledge } from '../src/knowledge.js';
import { MockLlmProvider } from '../src/llmProvider.js';

function makeAgent(client: MockLlmProvider, options: AgentOptions = {}): SmoothAgent {
    return new SmoothAgent(client, options);
}

describe('InMemoryKnowledge', () => {
    it('ranks by token overlap', () => {
        const kb = new InMemoryKnowledge();
        kb.ingest('The return window is 17 days from delivery.', 'returns.md');
        kb.ingest('Gift wrapping costs 4.99 per item.', 'wrapping.md');
        const hits = kb.query('what is the return window?', 1);
        expect(hits).toHaveLength(1);
        expect(hits[0].content).toContain('17 days');
    });
});

describe('SmoothAgent', () => {
    it('stops after one call on a text reply', async () => {
        const client = new MockLlmProvider().pushText('the answer is 42');
        const agent = makeAgent(client, { instructions: 'be helpful' });
        const result = await agent.run('what is the answer?');
        expect(result.text).toBe('the answer is 42');
        expect(result.iterations).toBe(1);
        expect(result.toolCalls).toBe(0);
    });

    it('runs a tool then finishes', async () => {
        const echo: Tool = {
            name: 'echo',
            description: 'Echoes input back',
            parameters: { type: 'object', properties: { text: { type: 'string' } }, required: ['text'] },
            execute: async (args) => String(args.text ?? ''),
        };
        const client = new MockLlmProvider();
        client.pushToolCall('call-1', 'echo', '{"text": "hello tools"}').pushText('done');
        const agent = makeAgent(client, { tools: [echo] });
        const result = await agent.run('use echo');
        expect(result.text).toBe('done');
        expect(result.toolCalls).toBe(1);
        // The tool result was fed back as a tool-role message before the final call.
        const secondCallMessages = client.calls[1].messages;
        expect(secondCallMessages.some((m) => m.role === 'tool' && m.content === 'hello tools')).toBe(true);
    });

    it('injects knowledge into the system prompt', async () => {
        const kb = new InMemoryKnowledge();
        kb.ingest('The return window is exactly 17 days from delivery.', 'returns.md');
        const client = new MockLlmProvider().pushText('17 days');
        const agent = makeAgent(client, { instructions: 'support agent', knowledge: kb });
        await agent.run('how many days to return?');
        const messages = client.calls[0].messages;
        expect(messages[0].role).toBe('system');
        expect(messages[0].content).toContain('17 days');
    });
});

describe('effectiveMaxTokens (model-output ceiling clamp, EPIC th-1cc9fa)', () => {
    it('clamps down when the ceiling is below the budget', () => {
        expect(effectiveMaxTokens(32768, 8192)).toBe(8192);
    });

    it('passes the budget through when the ceiling is >= the budget', () => {
        expect(effectiveMaxTokens(512, 8192)).toBe(512);
        expect(effectiveMaxTokens(8192, 8192)).toBe(8192);
    });

    it('passes the budget through when there is no ceiling (undefined or non-positive)', () => {
        expect(effectiveMaxTokens(8192, undefined)).toBe(8192);
        expect(effectiveMaxTokens(8192, 0)).toBe(8192);
        expect(effectiveMaxTokens(8192, -5)).toBe(8192);
    });

    it('never returns 0 even for a tiny ceiling', () => {
        expect(effectiveMaxTokens(8192, 1)).toBe(1);
        // A pathological budget of 0 with a positive ceiling still floors at 1.
        expect(effectiveMaxTokens(0, 4)).toBe(1);
    });
});

describe('SmoothAgent max_tokens clamp', () => {
    it('sends the clamped max_tokens on the non-streaming request when a ceiling is set', async () => {
        const client = new MockLlmProvider().pushText('ok');
        const agent = makeAgent(client, { maxTokens: 32768, modelMaxOutput: 8192 });
        await agent.run('hi');
        expect(client.calls[0].body.max_tokens).toBe(8192);
    });

    it('sends the unclamped budget when no ceiling is set', async () => {
        const client = new MockLlmProvider().pushText('ok');
        const agent = makeAgent(client, { maxTokens: 32768 });
        await agent.run('hi');
        expect(client.calls[0].body.max_tokens).toBe(32768);
    });

    it('sends the clamped max_tokens on the streaming request when a ceiling is set', async () => {
        const client = new MockLlmProvider().pushText('ok');
        const agent = makeAgent(client, { maxTokens: 32768, modelMaxOutput: 8192 });
        // Drive runStream to completion so the request body is recorded.
        for await (const _ of agent.runStream('hi')) {
            // drain
        }
        expect(client.calls[0].body.max_tokens).toBe(8192);
    });
});
